modules/java-modules/hdfs/src/main/java/org/syslog_ng/hdfs/HdfsDestination.java

```diff
@@ -60,11 +66,97 @@
 
     HashMap<String, HdfsFile> openedFiles;
 
+    private Timer timeReap;
+
+    private Lock lock;
+
+    class TimeReap extends TimerTask {
+        private Map<String, HdfsFile> files;
+        public TimeReap(Map<String, HdfsFile> files) {
+            logger.trace("Create TimeReap");
+            this.files = files;
+        }
+
+        private boolean closeFile(HdfsFile hdfsfile) {
+           logger.info("Close file" + hdfsfile.getPath());
+
+           try {
+               hdfsfile.flush();
+               hdfsfile.close();
+           } catch (IOException e) {
+               printStackTrace(e);
+               return false;
+           }
+
+           archiveFile(hdfsfile);
+
+           return true;
+        }
+
+        void closeExpiredFiles() {
+           for (Map.Entry<String, HdfsFile> entry : files.entrySet()) {
+               HdfsFile file = entry.getValue();
+               long current = file.timeSinceLastWrite();
+
+               if (current>0) {
+                  logger.trace("time-reap file " + file.getPath() + " since last write " + current);
+               }
+
+               if (current >= options.getTimeReap())
+                  closeFile(file);
+           }
+        }
+
+        long caculateNextWakeUp() {
+           if (files.isEmpty())
+              return 0;
+
+           long nextTimerExpires = -1;
+           for (Map.Entry<String, HdfsFile> entry : files.entrySet()) {
+               long current = entry.getValue().timeSinceLastWrite();
+
+               if (current > nextTimerExpires || nextTimerExpires == -1) {
+                  nextTimerExpires = current;
+               }
+           }
+
+           if (nextTimerExpires <= 0)
+              return 0;
+
+           nextTimerExpires = Math.abs(options.getTimeReap() - nextTimerExpires);
+           logger.debug("Next schedule in " + nextTimerExpires);
+
+           return nextTimerExpires;
+        }
+
+        @Override
+        public void run() {
+           logger.debug("time-reap run");
+           lock.lock();
+           try {
+           closeExpiredFiles();
+
+           long nextTimerExpires = caculateNextWakeUp();
+
+           timeReap.cancel();
```

Comment:
I don't think we need to cancel other tasks, since as I see it we create a new timer every time we need to reschedule.
The other schedule logic is in `send()` method, which can only schedule a timer if there is none.
We can't have multiple tasks with the same timer.

I'm not a java expert, but I saw that you can reuse a timer, just add a new TimerTask instance.
Otherwise I would use `Timer.purge()` as well to be able to gc the cancelled tasks.
See [Timer](https://docs.oracle.com/javase/8/docs/api/java/util/Timer.html#cancel())

Reply:
thanks.